apps/playwright-browser-tunnel/src/PlaywrightBrowserTunnel.ts

```diff
@@ -0,0 +1,524 @@
+
+interface IHandshake {
+  action: 'handshake';
+  browserName: BrowserNames;
+  launchOptions: LaunchOptions;
+  playwrightVersion: semver.SemVer;
+}
+
+type ITunnelMode = 'poll-connection' | 'wait-for-incoming-connection';
+
+/**
+ * Options for configuring a {@link PlaywrightTunnel} instance.
+ * @alpha
+ */
+export type IPlaywrightTunnelOptions = {
+  terminal: ITerminal;
+  onStatusChange: (status: TunnelStatus) => void;
+  tmpPath: string;
+} & (
+  | {
+      mode: 'poll-connection';
+      wsEndpoint: string;
+    }
+  | {
+      mode: 'wait-for-incoming-connection';
+      listenPort: number;
+    }
+);
+
+interface IBrowserServerProxy {
+  browserServer: BrowserServer;
+  client: WebSocket;
+}
+
+type ISupportedBrowsers = 'chromium' | 'firefox' | 'webkit';
+
+async function sleep(ms: number): Promise<void> {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+/**
+ * Hosts a Playwright browser server and forwards traffic over a WebSocket tunnel.
+ * @alpha
+ */
+export class PlaywrightTunnel {
+  private readonly _terminal: ITerminal;
+  private readonly _onStatusChange: (status: TunnelStatus) => void;
+  private readonly _playwrightBrowsersInstalled: Set<string> = new Set();
+  private _status: TunnelStatus = 'stopped';
+  private _initWsPromise?: Promise<WebSocket>;
+  private _keepRunning: boolean = false;
+  private _ws?: WebSocket;
+  private _mode: ITunnelMode;
+  private readonly _wsEndpoint?: string;
+  private readonly _listenPort?: number;
+  private readonly _tmpPath: string;
+
+  public constructor(options: IPlaywrightTunnelOptions) {
+    const { mode, terminal, onStatusChange, tmpPath } = options;
+
+    if (mode === 'poll-connection') {
+      if (!options.wsEndpoint) {
+        throw new Error('wsEndpoint is required for poll-connection mode');
+      }
+      this._wsEndpoint = options.wsEndpoint;
+    } else if (mode === 'wait-for-incoming-connection') {
+      if (options.listenPort === undefined) {
+        throw new Error('listenPort is required for wait-for-incoming-connection mode');
+      }
+      this._listenPort = options.listenPort;
+    } else {
+      throw new Error(`Invalid mode: ${mode}`);
+    }
+
+    this._mode = mode;
+    this._terminal = terminal;
+    this._onStatusChange = onStatusChange;
+    this._tmpPath = tmpPath;
+  }
+
+  public get status(): TunnelStatus {
+    return this._status;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  private set status(newStatus: TunnelStatus) {
+    this._status = newStatus;
+    this._onStatusChange(newStatus);
+  }
+
+  public async waitForCloseAsync(): Promise<void> {
+    const terminal: ITerminal = this._terminal;
+    await new Promise<void>((resolve) => {
+      void this._initWsPromise?.then((ws) => {
+        ws.on('close', () => {
+          terminal.writeLine('WebSocket connection closed. resolving init promise.');
+          this._initWsPromise = undefined;
+          resolve();
+        });
+      });
```

Comment:
This construction will hang if `this._initWsPromise` is not defined. Recommend assigning the promise to a variable outside of the `new Promise` to remove the optional chaining.

Also, we can use `once(ws, 'close')` (imported from `node:events`) instead of `new Promise`